csharp/src/Ice/OutgoingResponseFrame.cs

```diff
@@ -9,83 +9,119 @@
+        /// <param name="encoding">The encoding for the frame payload.</param>
+        public static OutgoingResponseFrame WithVoidReturnValue(Encoding encoding)
+            => new OutgoingResponseFrame(encoding, true);
 
-        /// <summary>Creates a new outgoing response frame. It's a partial frame, and its payload needs to be
-        /// written later on with StartReturnValue/EndReturnValue.</summary>
-        /// <param name="current">The current parameter holds decoded header data and other information about the
-        /// request for which this method creates a response.</param>
-        public OutgoingResponseFrame(Current current) : base(current.Adapter.Communicator, Ice1Definitions.Encoding)
+        public static OutgoingResponseFrame WithReturnValue<T>(Encoding encoding, FormatType? format,
+            in T value, OutputStreamWriter<T> writer)
         {
-            RequestId = current.RequestId;
-            _payloadEncoding = current.Encoding;
+            var response = new OutgoingResponseFrame(encoding, false);
+            byte[] buffer = new byte[256];
+            buffer[0] = (byte)ReplyStatus.OK;
+            response.Data.Add(buffer);
+            var ostr = new OutputStream(encoding, response.Data, new OutputStream.Position(0, 1), format);
+            writer(ostr, value);
+            ostr.Save();
+            response.Finish();
+            return response;
+        }
 
-            WriteSpan(Ice1Definitions.ReplyHeader.AsSpan());
-            WriteInt(RequestId);
+        public static OutgoingResponseFrame WithReturnValue<T>(Encoding encoding, FormatType? format, in T value,
+            OutputStreamStructWriter<T> writer) where T : struct
+        {
+            var response = new OutgoingResponseFrame(encoding, false);
+            byte[] buffer = new byte[256];
+            buffer[0] = (byte)ReplyStatus.OK;
+            response.Data.Add(buffer);
+            var ostr = new OutputStream(encoding, response.Data, new OutputStream.Position(0, 1), format);
+            writer(ostr, value);
+            ostr.Save();
+            response.Finish();
+            return response;
         }
 
         /// <summary>Creates a new outgoing response frame with the given reply status and payload.</summary>
-        /// <param name="current">The current parameter holds decoded header data and other information about the
-        /// request for which this method creates a response.</param>
-        /// <param name="replyStatus">The reply status.</param>
+        /// <param name="encoding">The encoding for the frame payload.</param>
         /// <param name="payload">The payload for this response frame.</param>
         // TODO: add parameter such as "bool assumeOwnership" once we add memory pooling.
-        public OutgoingResponseFrame(Current current, ReplyStatus replyStatus, ArraySegment<byte> payload)
-            : this(current)
+        public OutgoingResponseFrame(Encoding encoding, ArraySegment<byte> payload) :
```

Comment:
+ other ctor with IList<ArraySegment>?

Reply:
>other ctor with IList?
I added a TODO for this, it is not clear if we need both or just the one with a segment list.